[tools/depends] bump expat 2.4.1

fuzzard opened the pull request:
## Description
bump expat 2.4.1

## Motivation and context
dependency bump

@wsnipex another one thanks

A TODO here would be to switch over from autoconf to cmake for building expat. 

@Paxxi not sure where windows is regarding versions for expat, but noticed in the changelogs of expat 2.2.8 they renamed their cmake options
https://github.com/libexpat/libexpat/blob/a28238bdeebc087071777001245df1876a11f5ee/expat/Changes#L262-L279
They also deprecated Visual Studio <=14.0/2015 in 2.3.0 - https://github.com/libexpat/libexpat/blob/a28238bdeebc087071777001245df1876a11f5ee/expat/Changes#L114
May or may not be useful to you, just something i spotted.

## How has this been tested?
build osx

## What is the effect on users?
N/A

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[tools/depends][native] bump expat 2.4.1

Commit:
[tools/depends][target] bump expat 2.4.1

fix python module build with system expat for elementtree and pyexpat causing duplicate
symbols.
with-system-* flags dont actually do anything for us, as we dont (cant) use the auto
detection for python modules, and we instead rely on the Modules.setup to build everything

fuzzard commented:
Going to need some advice on this one.

```
duplicate symbol '_testingAccountingGetCountBytesIndirect' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_testingAccountingGetCountBytesDirect' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_XML_SetBillionLaughsAttackProtectionMaximumAmplification' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_unsignedCharToPrintable' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
duplicate symbol '_XML_SetBillionLaughsAttackProtectionActivationThreshold' in:
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libpython3.9.a(xmlparse.o)
    /Users/Shared/xbmc-depends/macosx12.0_x86_64-target-debug/lib/libexpat.a(xmlparse.o)
ld: 5 duplicate symbols for architecture x86_64
```

All of these symbols were introduced in expat 2.4.0+, and essentially we are getting duplicate symbols from static build of python and then our use of expat in the kodi binary.

My solution was to introduce a dependency (expat-python) and then have expat then depend on it to insure the python build and use of python expat uses one set of symbols, and then the rest use renamed symbols.
Seems messy, but cant come up with a better way. Ideas greatly appreciated.

wsnipex commented:
Are those warnings or errors? I'd assume ld issuing a warning, but silently using the first lib listed to link against.

Paxxi commented:
Doesn't python want expat as a shared library _expat.pyd for the module?

fuzzard commented:
> Are those warnings or errors? I'd assume ld issuing a warning, but silently using the first lib listed to link against.

They are errors on clang
```
ld: 6 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

As for the shared library, everything in the depends build for python modules are built static. So i guess if that was the required way, would have to rework everything in python

wsnipex commented:
is the LDFLAG "-all_load" enabled? That would explain the error

fuzzard commented:
Not that i can find.

fuzzard commented:
> Doesn't python want expat as a shared library _expat.pyd for the module?

So with a quick workup, the reason python modules are handled the way they are currently (statically built into libpython.x.x.a) is android doesnt play nice with loading from default python locations (lib/pythonx.x/lib-dynload).
Would need to add more patches to make it work. 
I dont think at this time its worth going through that hassle

garbear commented on tools/depends/target/Makefile:
```diff
@@ -15,7 +15,8 @@ DEPENDS = \
 	libxslt ffmpeg crossguid libudfread \
-	libdvdread libdvdnav libdvdcss p8-platform flatbuffers dav1d xz
+	libdvdread libdvdnav libdvdcss p8-platform flatbuffers dav1d xz \
+  expat-python
```
do we care about a little misalignment?

fuzzard replied:
I'll fix that up in the rebase if the approach is deemed acceptable

garbear replied:
I'm having flashbacks of all the times I've fixed rebase conflicts in this exact stanza. Please let's sort it and break into individual lines.

Commit:
[tools/depends] makefile fixups

Spotted by Rechi, some fixups to garbears PR https://github.com/xbmc/xbmc/pull/20582

fuzzard commented:
Heres my latest attempt. once again, opinions very much wanted.

This time, bring in a patch to set functions that cause duplicate symbols to be static functions.

The following 3 symbols had static removed due to test suite. For our purposes, they are internal, and we dont run test suite, so can either static or inline them.

- testingAccountingGetCountBytesDirect
- testingAccountingGetCountBytesIndirect
- unsignedCharToPrintable

_INTERNAL_trim_to_complete_utf8_characters is also an internal only function, and the symbol was unexposed in https://github.com/libexpat/libexpat/pull/457

That leaves the following 2 that ive just set to static.

- XML_SetBillionLaughsAttackProtectionMaximumAmplification
- XML_SetBillionLaughsAttackProtectionActivationThreshold

Local testing gets building working. In theory i think it should be ok, but would like other opinions.

fuzzard commented:
Ok, round 3... haha.

The alternate i can think of, and as i now implement it, i think its the way to go.

Currently for target python3 we use system supplied expat (ie tools/depends/target/expat lib). Python supplies expat internally as an option and currently it is up to date with expat 2.41. As part of their pyexpat module, they redefine expat declarations in  https://github.com/python/cpython/blob/main/Modules/expat/pyexpatns.h

This has not been updated to support the new expat functions. If we simply patch this to include the "new" expat functions, and link with this instead of system expat, we then get different symbols in libpython3.9.a and libexpat.a when kodi is linked.

problem solved, patch in python is far simpler to maintain than the previous expat patch. The only potential downside is no longer using system expat, therefore if cpython upstream are sloppy with keeping it up to date, it may fall behind. I think at this point in time its a tradeoff worth taking.

Once again, let me know preferences for which approach

wsnipex commented:
Even though it's not optimal to use internal expat in python, it does look like a sane solution for the time being for finally get expat updated.
